src/helpers/figures/charts/bar_chart.ts

```diff
@@ -110,47 +116,47 @@ export class BarChart extends AbstractChart {
     return {
       background: this.background,
       title: this.title,
-      range: this.dataSets.map((ds: DataSet) =>
-        this.getters.getRangeString(ds.dataRange, this.sheetId)
-      ),
+      range: this.definition.dataSets,
       auxiliaryRange: this.labelRange
         ? this.getters.getRangeString(this.labelRange, this.sheetId)
         : undefined,
     };
   }
 
   copyForSheetId(sheetId: UID): BarChart {
-    const dataSets = copyDataSetsWithNewSheetId(this.sheetId, sheetId, this.dataSets);
+    const dataSets = copyDataSetsWithNewSheetId(this.sheetId, sheetId, this.getDataSets()).map(
+      (ds: DataSet) => this.getters.getRangeString(ds.dataRange, sheetId)
+    );
     const labelRange = copyLabelRangeWithNewSheetId(this.sheetId, sheetId, this.labelRange);
     const definition = this.getDefinitionWithSpecificDataSets(dataSets, labelRange, sheetId);
     return new BarChart(definition, sheetId, this.getters);
   }
 
   copyInSheetId(sheetId: UID): BarChart {
-    const definition = this.getDefinitionWithSpecificDataSets(
-      this.dataSets,
-      this.labelRange,
-      sheetId
+    const dataSets = this.getDataSets().map((ds: DataSet) =>
+      this.getters.getRangeString(ds.dataRange, sheetId)
     );
+    const definition = this.getDefinitionWithSpecificDataSets(dataSets, this.labelRange, sheetId);
     return new BarChart(definition, sheetId, this.getters);
   }
 
   getDefinition(): BarChartDefinition {
-    return this.getDefinitionWithSpecificDataSets(this.dataSets, this.labelRange);
+    const dataSets = this.getDataSets(false).map((ds: DataSet) =>
+      this.getters.getRangeString(ds.dataRange, this.sheetId)
+    );
+    return this.getDefinitionWithSpecificDataSets(dataSets, this.labelRange);
```

Comment:
Isn't that the same as just returning this.definition that was given to the constructor ? The goal was to return the raw definiton without truncating the datasets right ? If yes we can remove the ugly boolean argument in this.getDataSets() and make it a getter (or init at constructor)

Reply:
No it's not the same as this._definition. We have to go through through this.getDataSets to "clean" the datasets (removing the sheetName in the range string if it's the same as the chart sheet name, removing invalid datasets, ...)